addons/api/addon/serializers/storage-bucket.js

```diff
@@ -4,29 +4,62 @@
  */
 
 import ApplicationSerializer from './application';
+import { TYPE_CREDENTIAL_DYNAMIC } from '../models/storage-bucket';
 
+const fieldsByCredentialType = {
+  static: [
+    'access_key_id',
+    'secret_access_key',
+    'region',
+    'disable_credential_rotation',
+  ],
+  dynamic: [
+    'role_arn',
+    'role_external_id',
+    'role_session_name',
+    'role_tags',
+    'disable_credential_rotation',
+    'region',
+  ],
+};
 export default class StorageBucketSerializer extends ApplicationSerializer {
   serialize(snapshot) {
     const serialized = super.serialize(...arguments);
-    //Dynamic credentials do not allow credential rotation, and the API requires this field to be
-    //sent along with the dynamic credential fields
-    if (serialized.attributes.role_arn) {
-      serialized.attributes.disable_credential_rotation = true;
-    }
+    const { credentialType } = snapshot.record;
     if (!snapshot.record.isNew) {
       delete serialized.bucket_name;
       delete serialized.bucket_prefix;
     }
+
+    if (credentialType === TYPE_CREDENTIAL_DYNAMIC) {
+      serialized.attributes.disable_credential_rotation = true;
+    }
     return serialized;
   }
 
+  serializeAttribute(snapshot, json, key, attribute) {
+    const value = super.serializeAttribute(...arguments);
+    const { credentialType } = snapshot.record;
+    const { options } = attribute;
+    // This deletes any fields that don't belong to the record's credential type
+    if (options.isNestedAttribute && json.attributes) {
```

Comment:
Does this work correctly for `access_key_id` and `secret_access_key`? They're only annotated as `isNestedSecret` so wouldn't we skip over them?